### PR TITLE
[mono] PR tracking the 2017-04 bump.

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

**Comment by jonpryor** [comment]
```
stderr: error: no such remote ref 3e5c484fe72f9d681206312d33a15f0f93e7dde7
14:58:25 Fetched in submodule path 'external/Java.Interop', but it did not contain 3e5c484fe72f9d681206312d33a15f0f93e7dde7. Direct fetching of that commit failed.
```

That's probably because 3e5c484f is kumpera/Java.Interop/3e5c484f, *not* xamarin/Java.Interop/3e5c484f.

If you create a branch from xamarin/Java.Interop and not kumpera/Java.Interop, this will presumably work. *Or*, if you "fix" the `external/Java.Interop` ref to instead refer to `kumpera/Java.Interop`, I imagine that would work.

...or we accept https://github.com/xamarin/java.interop/pull/141 and then you can bump "normally."

That Cecil bump looks "big."

**Comment by kumpera** [comment]
build

**Comment by radical** [comment]
Can we rebuild this? The CI build is failing with
 ```
15:43:47  > git fetch --tags --progress git@github.com:xamarin/xamarin-android +refs/pull/*:refs/remotes/origin/pr/*
15:43:49 ERROR: Error fetching remote repo 'origin'
```

**Comment by radical** [comment]
build

**Comment by radical** [comment]
```17:15:29 /mnt/jenkins/workspace/xamarin-anroid-linux-pr-builder/xamarin-android/build-tools/dependencies/dependencies.mdproj (default targets) ->
17:15:29 /mnt/jenkins/workspace/xamarin-anroid-linux-pr-builder/xamarin-android/build-tools/scripts/RequiredPrograms.targets (CheckForRequiredPrograms target) ->
17:15:29 
17:15:29 	/mnt/jenkins/workspace/xamarin-anroid-linux-pr-builder/xamarin-android/build-tools/scripts/RequiredPrograms.targets: error : Missing dependency detected. For Linux we do not know how to install program `mono`, version >= 5.2.0.```

Could somebody look at this?

**Comment by jonpryor** [comment]
@directhex: wrt @radical's [question](https://github.com/xamarin/xamarin-android/pull/551#issuecomment-297781802), is there some preferred mechanism to convince "the" Linux builder to install Mono 5.2 instead of (presumably) the current 4.9?

I think we're going to need some mechanism of specifying which mono version to install, and I'm not sure what that should be. Could we e.g. add a Linux lane specific to this PR so that it installs the correct Mono version until merged on master? Something else?

**Comment by directhex** [comment]
@jonpryor the preferred mechanism crumbles a bit when the request is for a version which isn't in alpha yet.

I already added a mechanism to make it possible to add one-off builds as repositories, what I don't have is any package-related one-off builds of the 5.2 branch. I'll see what I can do.

**Comment by kumpera** [comment]
We ignore linux CI results when testing newer mono. Looks like everything is good now!

**Comment by radical** [comment]
I was looking at this because of https://bugzilla.xamarin.com/show_bug.cgi?id=55287 . Can the bug be closed now?

**Comment by directhex** [comment]
Is Mono master (5.3.x) usable, or does it need to be 5.2?

**Comment by kumpera** [comment]
@directhex I think 5.4 would do it, but it's not strictly needed.

I think this PR is ready for merging as is once the linker conflict is resolved.

**Comment by kumpera** [comment]
@directhex ugg, forget that, we'll need those packages to move this forward.

**Comment by jonpryor** [comment]
The current plans to address the linker conflict are for @radekdoulik to *remove* `xamarin-android/external/linker` and instead rely on mono's linker checkout. I believe this is intended to land this week.

**Comment by directhex** [comment]
A `prealpha` repo now exists, with 5.2 packages. What am I doing with it?

**Comment by jonpryor** [comment]
@directhex: Do what you did with PR #445: when the Jenkins+Linux instance is started, have it install the `pre alpha` repo with 5.2 packages, then continue building.

This should be done for the `xamarin-andorid-linux` and `xamarin-android-linux-pr-builder` jobs:

* <https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-linux/>
* <https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-anroid-linux-pr-builder/>

**Comment by directhex** [comment]
OK, it's not *finished* building, but it's actually starting now, so my work here is done.

**Comment by radekdoulik** [comment]
@marek-safar we are missing this commit here for XA: https://github.com/mono/linker/commit/6e4e50c9cf1292271bc5e40b7fd0ed970e39196b

would you mind if I create a new branch on linker named mono-2017-04, cheery-pick the commit and let mono/2017-04 use the linker branch?

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.csproj [hunk 1]
```diff
@@ -343,6 +343,12 @@
     <Compile Include="$(LinkerSourceFullPath)\linker\Mono.Linker.Steps\MarkStep.cs">
       <Link>Linker\Mono.Linker.Steps\MarkStep.cs</Link>
     </Compile>
+    <Compile Include="$(LinkerSourceFullPath)\linker\Mono.Linker\LoadException.cs">
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.csproj]: This is currently [breaking the build](https://jenkins.mono-project.com/job/xamarin-android-msbuild-pr-builder/445/console):

```
13:27:12   CSC : error CS2001: Source file '/Users/builder/jenkins/workspace/xamarin-android-msbuild-pr-builder/external/linker/linker/Mono.Linker/LoadException.cs' could not be found.
```

[linker/81007dfb does not contain a `LoadException.cs` there](https://github.com/mono/linker/tree/81007dfb0c48c1908d7540571d76fc7107128714/linker/Mono.Linker).

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.csproj [hunk 2]
```diff
@@ -343,6 +343,12 @@
...
+      <Link>Linker\Mono.Linker\LoadException.cs</Link>
+    </Compile>
+    <Compile Include="$(LinkerSourceFullPath)\linker\Mono.Linker\MarkException.cs">
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Build.Tasks.csproj]: ...ditto `MarkException.cs`

**Pull request merged**